test/e2e_node/oomkiller_linux_test.go

```diff
@@ -120,18 +143,34 @@ func runOomKillerTest(f *framework.Framework, testCase testCase, kubeReservedMem
 			e2epod.NewPodClient(f).Create(context.TODO(), testCase.podSpec)
 		})
 
-		ginkgo.It("The containers terminated by OOM killer should have the reason set to OOMKilled", func() {
-			ginkgo.By("Waiting for the pod to be failed")
-			err := e2epod.WaitForPodTerminatedInNamespace(context.TODO(), f.ClientSet, testCase.podSpec.Name, "", f.Namespace.Name)
-			framework.ExpectNoError(err, "Failed waiting for pod to terminate, %s/%s", f.Namespace.Name, testCase.podSpec.Name)
+		if testCase.expectPodRunning {
+			ginkgo.It("The containers should not be OOMKilled", func() {
+				gomega.Consistently(context.TODO(), func(ctx context.Context) error {
+					pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, testCase.podSpec.Name, metav1.GetOptions{})
+					if err != nil {
+						return fmt.Errorf("expected the pod %s to exist: %w", pod.Name, err)
+					}
+					phase := pod.Status.Phase
+					if phase != v1.PodRunning && phase != v1.PodPending && phase != v1.PodSucceeded {
```

Comment:
I understand that we'd like to get this merged before the code freeze, but I just had a few concerns I wanted to comment on.

With the current way the test is written, the time spent in the `PodPending` state is also included, so it may not be ideal for checking whether an OOMKill occurs. By rewriting it as shown below, I think we can remove the PodPending state condition and verify that it stays in either the `PodRunning` or `PodSucceeded` state continuously for 10 seconds.

```diff
diff --git a/test/e2e_node/oomkiller_linux_test.go b/test/e2e_node/oomkiller_linux_test.go
index ad2bff26c17..a9e2c4fe8f5 100644
--- a/test/e2e_node/oomkiller_linux_test.go
+++ b/test/e2e_node/oomkiller_linux_test.go
@@ -145,14 +145,17 @@ func runOomKillerTest(f *framework.Framework, testCase testCase, kubeReservedMem

                if testCase.expectPodRunning {
                        ginkgo.It("The containers should not be OOMKilled", func() {
+                               err := e2epod.WaitForPodsRunning(context.TODO(), f.ClientSet, f.Namespace.Name, 1, framework.PodStartTimeout)
+                               framework.ExpectNoError(err, "Failed waiting for pod to be running state, %s/%s", f.Namespace.Name, testCase.podSpec.Name)
+
                                gomega.Consistently(context.TODO(), func(ctx context.Context) error {
                                        pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, testCase.podSpec.Name, metav1.GetOptions{})
                                        if err != nil {
                                                return fmt.Errorf("expected the pod
%s to exist: %w", pod.Name, err)
                                        }
                                        phase := pod.Status.Phase
-                                       if phase != v1.PodRunning && phase != v1.PodPending && phase != v1.PodSucceeded {
-                                               return fmt.Errorf("pod %s: unexpected status %s, expected status: %s, %s or %s", pod.Name, pod.Status.Phase, v1.PodRunning, v1.PodPending, v1.PodSucceeded)
+                                       if phase != v1.PodRunning && phase != v1.PodSucceeded {
+                                               return fmt.Errorf("pod %s: unexpected status %s, expected status: %s or %s", pod.Name, pod.Status.Phase, v1.PodRunning, v1.PodSucceeded)
                                        }
                                        return nil
                                }, 10*time.Second, f.Timeouts.Poll).Should(gomega.BeNil())
```

Is transitioning to the `PodSucceeded` state the expected behavior? In the case of a single-process OOM Kill, I think the process exceeding the resource limit will be killed, but the Pod itself won't stop. Could this be an issue with the way the stress command is written in the Pod spec?

Reply:
> With the current way the test is written, the time spent in the PodPending state is also included, so it may not be ideal for checking whether an OOMKill occurs. By rewriting it as shown below, I think we can remove the PodPending state condition and verify that it stays in either the PodRunning or PodSucceeded state continuously for 10 seconds.

Make sense. I'll address it.

> Is transitioning to the PodSucceeded state the expected behavior? In the case of a single-process OOM Kill, I think the process exceeding the resource limit will be killed, but the Pod itself won't stop. Could this be an issue with the way the stress command is written in the Pod spec?

I don't know if this is desirable. However, it should behave the same as cgroup v1, so I think the purpose of this PR will be achieved.

I think it is difficult to control which process is OOMed in the test. So, we should allow Success even if you tweak the command. In that case, I assume that the long sleep command was killed by OOM.

Reply:
Addressed